pkg/operator/ceph/csi/secrets.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2019 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"fmt"
+
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/ceph/config/keyring"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	csiKeyringRBDProvisionerUsername = "client.csi-rbd-provisioner"
+	csiKeyringRBDNodeUsername        = "client.csi-rbd-node"
+	csiRBDNodeSecret                 = "rook-csi-rbd-node"
+	csiRBDProvisionerSecret          = "rook-csi-rbd-provisioner"
+)
+
+const (
+	csiKeyringCephFSProvisionerUsername = "client.csi-cephfs-provisioner"
+	csiKeyringCephFSNodeUsername        = "client.csi-cephfs-node"
+	csiCephFSNodeSecret                 = "rook-csi-cephfs-node"
+	csiCephFSProvisionerSecret          = "rook-csi-cephfs-provisioner"
+)
+
+func createCSIKeyringRBDNode(s *keyring.SecretStore) (string, error) {
+	key, err := s.GenerateKey(csiKeyringRBDNodeUsername, cephCSIKeyringRBDNodeCaps())
+	if err != nil {
+		return "", err
+	}
+
+	return key, nil
+}
+
+func createCSIKeyringRBDProvisioner(s *keyring.SecretStore) (string, error) {
+	key, err := s.GenerateKey(csiKeyringRBDProvisionerUsername, cephCSIKeyringRBDProvisionerCaps())
+	if err != nil {
+		return "", err
+	}
+
+	return key, nil
+}
+
+func createCSIKeyringCephFSNode(s *keyring.SecretStore) (string, error) {
+	key, err := s.GenerateKey(csiKeyringCephFSNodeUsername, cephCSIKeyringCephFSNodeCaps())
+	if err != nil {
+		return "", err
+	}
+
+	return key, nil
+}
+
+func createCSIKeyringCephFSProvisioner(s *keyring.SecretStore) (string, error) {
+	key, err := s.GenerateKey(csiKeyringCephFSProvisionerUsername, cephCSIKeyringCephFSProvisionerCaps())
+	if err != nil {
+		return "", err
+	}
+
+	return key, nil
+}
+
+func cephCSIKeyringRBDNodeCaps() []string {
+	return []string{
+		"mon", "profile rbd",
+		"osd", "profile rbd",
+	}
+}
+
+func cephCSIKeyringRBDProvisionerCaps() []string {
+	return []string{
+		"mon", "profile rbd",
+		"mgr", "allow rw",
+		"osd", "profile rbd",
+	}
+}
+
+func cephCSIKeyringCephFSNodeCaps() []string {
+	return []string{
+		"mon", "allow r",
+		"osd", "allow rw tag cephfs *=*",
+		"mds", "allow rw",
```

Comment:
Need to append,
"mgr", "allow rw",

Reply:
I'm not sure this one is necessary if the application metadata fixes are added. Let's try reverting this.

Reply:
We still need the MGR caps to enable csi-cephfs-node pod to issue the `fs subvolume getpath` command 

I guess you're talking about  OSD caps of client.csi-cephfs-provisioner and client.csi-cephfs-node?

The OSD caps of `client.csi-cephfs-node` needs to be as is,
```
"osd", "allow rw tag cephfs *=*"
```
Note:  The `client.csi-cephfs-node` is used to fetch omapval from CephFS's metadata pool, and used to read/write directly from/to CephFS's data pool.

The OSD caps of client.csi-cephfs-provisioner can be changes to,
```
"osd", "allow rw tag cephfs metadata=*"
```
Note: we specify `metadata=*` as we need to allow `client.csi-cephfs-provisioner` to fetch omap kvs stored in CephFS's metadata pool